### PR TITLE
Fix blanket `# noqa` in astropy

### LINKED ISSUE #13702: Add pre-commit pygrep hook `python-check-blanket-noqa`
### Description

We have a lot of blanket ``# noqa`` (instead of e.g. ``# noqa: F401``) , which is bad practice. If we add the pre-commit hook https://github.com/pre-commit/pygrep-hooks, then no more blanket ``# noqa`` can be added.

### PR DESCRIPTION
**Pull request by WilliamJamieson**
### Description

This pull request fixes all of the blanket `# noqa` throughout `astropy`. Either by fixing the reason for the `# noqa`, removing it as it was unnecessary, or by adding the appropriate  violation code to be skipped.

Fixes #13702

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label. Codestyle issues can be fixed by the [bot](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit).
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by WilliamJamieson** [comment]
@nstarman or @pllim should I break this into multiple PRs?

**Comment by pllim** [comment]
For less than 300 lines, I think it can stay in one PR unless this drags on and you get a lot of conflicts, but I hope not.

File: astropy/coordinates/tests/test_pickle.py
Comment (nstarman) [inline astropy/coordinates/tests/test_pickle.py:67]: It would be nice to ignore this noqa, but ONLY in tests. Pytest fixtures do not play nicely with PEP8.

File: astropy/coordinates/tests/test_sky_coord.py
Comment (nstarman) [inline astropy/coordinates/tests/test_sky_coord.py:363]: No need for the comment anymore.

```suggestion
    assert sc1[0] == sc2[0]
```

File: astropy/cosmology/__init__.py
Comment (nstarman) [inline astropy/cosmology/__init__.py:15]: Oops. I had the intent, but not the correct execution 😆

File: astropy/table/tests/test_operations.py
Comment (nstarman) [inline astropy/table/tests/test_operations.py:833]: I'd restore a noqa to these types of tests. The ``isinstance`` replacement is less strict a comparison than the ``is``.

File: astropy/time/tests/test_basic.py
Comment (nstarman) [inline astropy/time/tests/test_basic.py:2384]: Same here.

**Review by nstarman — CHANGES_REQUESTED** [review]
Thanks @WilliamJamieson for going through all these files!

File: astropy/table/tests/test_operations.py
Comment (WilliamJamieson) [reply astropy/table/tests/test_operations.py]: There are a lot more of these issues scattered throughout `astropy` (we don't catch them because we don't acutally run our entire `flake8` config).

So instead of adding a bunch more `# noqa` flags I added a helper function which implements the "type is type" assertion (strangely `flake8` is fine with that one?). I did this in fb658d8, do you think this is better?

File: astropy/coordinates/tests/test_pickle.py
Comment (nstarman) [reply astropy/coordinates/tests/test_pickle.py:67]: Looking at https://github.com/pre-commit/pygrep-hooks/blob/main/.pre-commit-hooks.yaml, I don't think refinement is possible, so the noqa is fine.

File: astropy/tests/helper.py
Comment (nstarman) [inline astropy/tests/helper.py:481]: This file is actually public-ish API, so a more detailed doctring would be great.

```suggestion
    """Assert that two objects have exactly the same type.

    Parameters
    ------------
    a, b : object or type
        The two objects to compare. If either is a `type`,
         then compare directly to that type instead.

    Raises
    -------
    AssertionError
        If the type of ``a`` is not the type of ``b``.

    Examples
    ----------
    The following pass:

        >>> assert_type_is_type(int, int)
        >>> assert_type_is_type(1, int)
    """
```

File: astropy/table/tests/test_operations.py
Comment (nstarman) [reply astropy/table/tests/test_operations.py]: I think so! Might need a changelog entry for it

Comment (WilliamJamieson) [reply astropy/table/tests/test_operations.py]: Just added one!

**Review by nstarman — APPROVED** [review]
Thanks @WilliamJamieson for all the work! LGTM. I'll let other weigh in...

File: astropy/coordinates/tests/test_frames.py
Comment (pllim) [inline astropy/coordinates/tests/test_frames.py:894]: I would like coordinates maintainers to have a look at this one: @StuartLittlefair, @adrn, @eteq, @eerovaher, @mhvk (according to Team page)

File: astropy/io/ascii/cds.py
Comment (pllim) [inline astropy/io/ascii/cds.py:191]: Are you sure it won't complain about the lack of space after colon?

Comment (WilliamJamieson) [reply astropy/io/ascii/cds.py:191]: With our current flake8 configuration (from setup.cfg), there are no complaints from that whole file.

File: astropy/coordinates/tests/test_frames.py
Comment (mhvk) [reply astropy/coordinates/tests/test_frames.py:894]: I think the point here was indeed to show that `__eq__` returns a boolean (and not something like `"boo!"`, which would evaluate like true...), so I'd leave this one as it was (adding the relevant warning code).

File: astropy/coordinates/tests/test_sky_coord.py
Comment (mhvk) [inline astropy/coordinates/tests/test_sky_coord.py:363]: Same here

File: astropy/io/fits/tests/test_table.py
Comment (pllim) [inline astropy/io/fits/tests/test_table.py:1181]: @saimn , is it ok to remove `== True` here?

File: astropy/table/tests/test_column.py
Comment (pllim) [inline astropy/table/tests/test_column.py:799]: @taldcroft , is it okay to remove `== True` here?

File: astropy/io/ascii/tests/test_ecsv.py
Comment (mhvk) [inline astropy/io/ascii/tests/test_ecsv.py:969]: A bit sad to have to introduce a new function just to avoid code checkers!  I'd prefer the relevant `noqa`...

File: astropy/io/fits/hdu/image.py
Comment (mhvk) [inline astropy/io/fits/hdu/image.py:1085]: I think this is a functional change:
```
In [8]: 2 != True
Out[8]: True
```

File: astropy/io/fits/tests/test_table.py
Comment (mhvk) [inline astropy/io/fits/tests/test_table.py:1242]: Again, this may not right, it may well have been important to test that the array holds things equivalent to a bool

File: astropy/tests/helper.py
Comment (pllim) [reply astropy/tests/helper.py:500]: And @mhvk for the usage in time tests.

File: astropy/tests/helper.py
Comment (mhvk) [inline astropy/tests/helper.py:476]: As mentioned above, I find it a bit sad to have a function whose only reason is to avoid style checkers. Maybe we just do not enforce this particular item?

File: astropy/tests/helper.py
Comment (mhvk) [inline astropy/tests/helper.py:500]: This is not quite good enough:
```
In [9]: type (Time)
Out[9]: abc.ABCMeta
```
I think you actually want `isinstance(type(a), type)` :smiley_cat:

File: astropy/units/tests/test_units.py
Comment (pllim) [inline astropy/units/tests/test_units.py:174]: Why not just slap the `# noqa: E711` to be consistent with the test below?

File: docs/changes/tests/13754.feature.rst
Comment (pllim) [inline docs/changes/tests/13754.feature.rst:1]: Since this is exposed as "public" API here. Do we need a proper test to cover all the supported cases? The doctest snippet is pretty bare.

File: astropy/units/tests/test_units.py
Comment (mhvk) [inline astropy/units/tests/test_units.py:174]: I think here we are testing `unit.__ne__` so this should stay.

**Review by mhvk — COMMENTED** [review]
Thanks for doing this fairly thankless task!

I think the comparisons with `True`, `False`, and `None` should all stay -- they are meant to test `__eq__`, `__ne__`.

I'm also not very charmed with adding a function just for the sake of avoiding style checkers; let's just put the relevant number there (or allow it more generally).

**Comment by mhvk** [comment]
Seeing that @pllim and I had so many similar comments, perhaps the easiest way to resolve it is to make this PR one were we only make sure there are no bare `noqa`, either by removing them if they were not needed, or adding the specific warning.

Or do that in one commit and then the rest in separate ones (as already done for `assert_type_is_type` -- I think I'd drop that but can be convinced otherwise!)

**Comment by pllim** [comment]
Would definitely be easier to approve and merge faster if there is no significant code change besides `#noqa` stuff...

**Comment by WilliamJamieson** [comment]
Since these are all undocumented `# noqa` flags, it makes sense to attempt to fix whatever is causing the issue if possible.

Would it be better just to strip all the blanket `# noqa` flags instead.

File: astropy/coordinates/tests/test_frames.py
Comment (eerovaher) [reply astropy/coordinates/tests/test_frames.py:894]: If the intention is to test that the return values are indeed Booleans then the correct way to assert that would be
```python
assert (sc1[0] == sc2[0]) is True
assert (sc1[0] == sc2[0]) is False
```

File: astropy/io/ascii/cds.py
Comment (eerovaher) [reply astropy/io/ascii/cds.py:191]: It would be better to ensure all these lines are `black`-compliant.

File: astropy/coordinates/tests/test_frames.py
Comment (mhvk) [reply astropy/coordinates/tests/test_frames.py:894]: Yes, but note the comment in the original line: one gets a numpy True, not a python one...

Comment (taldcroft) [reply astropy/coordinates/tests/test_frames.py:894]: Would it make sense to write `assert (sc1[0] == sc2[0]) is np.bool_(True)`? Then the test is really insisting on getting the numpy True.

Comment (eerovaher) [reply astropy/coordinates/tests/test_frames.py:894]: We really should implement @taldcroft's suggestion because the current test does not distinguish between Python and `numpy` True:
```python
>>> import numpy as np
>>> np.bool_(True) == True
True
```

File: astropy/io/ascii/cds.py
Comment (eerovaher) [reply astropy/io/ascii/cds.py:191]: More lines being `black`-compliant already means the diffs `black` will create will be smaller.

File: astropy/coordinates/tests/test_frames.py
Comment (mhvk) [reply astropy/coordinates/tests/test_frames.py:894]: The test is really just to show that `__eq__` returns something like a boolean (i.e., not `sc` or `"boo!"` that evaluates as true). I think it is an implementation detail whether that is a numpy or python bool, so would keep the test as it was originally. Or do `assert isinstance(sc1[0] == sc2[0], (bool, np.bool_))` -- that makes the intent clearer.

File: astropy/table/tests/test_column.py
Comment (taldcroft) [reply astropy/table/tests/test_column.py:799]: Yes.

File: astropy/tests/helper.py
Comment (taldcroft) [reply astropy/tests/helper.py:476]: I think it is a relatively common mistake to check for types being equal instead of `isinstance`, so maybe leaving this check in place servers some value. If so then it seems we have two choices (1) use `# noqa (E721)` or (2) the new function. 

I'm a little worried about trying to be too clever here with the function as @mhvk pointed out. It's also going to be a bit obscure and it may be difficult to remember to use that. So maybe just using the explicit `noqa` in these cases will convey the intent most clearly?

Being "less clever" with the function as a drop-in replacement would essentially reduce it to `assert a is b`, which does seem a bit silly. With this code even the `# noqa` is not needed since the checker doesn't know `a` and `b` are types.

File: astropy/constants/__init__.py
Comment (astrofrog) [inline astropy/constants/__init__.py:30]: I'm curious if there would be a better way to deal with the warnings in ``__init__.py`` files since this is so common. There's nothing to fix here, so I wonder if we can somehow run flake8 in a way that it would automatically ignore F401/E402 when looking at ``__init__.py`` files?

File: astropy/coordinates/tests/test_pickle.py
Comment (astrofrog) [inline astropy/coordinates/tests/test_pickle.py:9]: Is this an unused import? Or is it because ``flake8`` doesn't understand that this is being used as a fixture?

File: astropy/coordinates/tests/test_sky_coord_velocities.py
Comment (astrofrog) [inline astropy/coordinates/tests/test_sky_coord_velocities.py:20]: Maybe cases like this mean that we should write ``skipif`` decorators using actual Python code instead of strings?

**Review by astrofrog — COMMENTED** [review]
I wonder if it would be worth splitting this effort into three:

* Removing any ``noqa`` that are no longer needed (this can probably be automated to avoid manual work). This should then be pretty uncontroversial.
* Finding a way to deal specifically with unused import warnings in ``__init__.py`` files since it's pretty tedious to have to clutter those files with so many ``noqa``s and is such a common thing.
* The rest

**Comment by nstarman** [comment]
> Finding a way to deal specifically with unused import warnings in __init__.py files since it's pretty tedious to have to clutter those files with so many noqas and is such a common thing.

My 2 cents... From my time spent static typing and compiling other libraries, I've found that star imports are better replaced by explicit imports. Pre-commit flake8 checks make sure the `__all__` in `__init__` stays synced with the module content.

**Comment by WilliamJamieson** [comment]
> My 2 cents... From my time spent static typing and compiling other libraries, I've found that star imports are better replaced by explicit imports. Pre-commit flake8 checks make sure the `__all__` in `__init__` stays synced with the module content.

Alternately, we could add a [`per-file-ignores`](https://flake8.pycqa.org/en/latest/user/options.html#cmdoption-flake8-per-file-ignores) entry to our flake8 config to ignore F401/E402 in `__init__.py` files.

**Comment by astrofrog** [comment]
I think @WilliamJamieson's suggestion would be best for now - and we can use wildcards so shouldn't be too hard.

**Comment by WilliamJamieson** [comment]
I like @astrofrog's suggestions on how to break this PR apart, so I think I'll create two separate PRs with the first removing all the unnecessary `# noqa` and the second forcing `flake8` and `pre-commit` to ignore the F401/E402 errors in `__init__.py`.

I will leave this PR open and try to keep it up to date so that we can discuss the more controversial `# noqa` changes here.

File: astropy/coordinates/tests/test_sky_coord_velocities.py
Comment (WilliamJamieson) [reply astropy/coordinates/tests/test_sky_coord_velocities.py:20]: I think this would be better if it would work.

File: astropy/coordinates/tests/test_pickle.py
Comment (WilliamJamieson) [reply astropy/coordinates/tests/test_pickle.py:9]: This is one of those unfortunate cases of `pytest` not being very "pythonic", which means, as you suggested, `flake8` doesn't understand its being used.

**Comment by WilliamJamieson** [comment]
I think I have addressed all the comments so far. I backed off on all the `type is type` changes and used `is np.bool_` in others.

I also removed all the `__init__.py` changes from this PR as those will be covered by #13781 when it merged. This means this will need to be rebased after #13781 is merged.

File: astropy/units/tests/test_units.py
Comment (WilliamJamieson) [reply astropy/units/tests/test_units.py:174]: Fixed by 757d1d9fd8c8b55ade21f6dc2d566c9ca5e8047f

File: astropy/tests/helper.py
Comment (WilliamJamieson) [reply astropy/tests/helper.py]: Based on @taldcroft's comment I decided to remove this and just use `# noqa: E721` instead.

File: astropy/io/fits/tests/test_table.py
Comment (WilliamJamieson) [reply astropy/io/fits/tests/test_table.py:1242]: Fixed by 757d1d9fd8c8b55ade21f6dc2d566c9ca5e8047f

File: astropy/io/ascii/tests/test_ecsv.py
Comment (WilliamJamieson) [reply astropy/io/ascii/tests/test_ecsv.py]: I undid this change in 757d1d9fd8c8b55ade21f6dc2d566c9ca5e8047f

File: astropy/units/tests/test_units.py
Comment (WilliamJamieson) [reply astropy/units/tests/test_units.py]: Fixed in 757d1d9fd8c8b55ade21f6dc2d566c9ca5e8047f

File: astropy/coordinates/tests/test_frames.py
Comment (WilliamJamieson) [reply astropy/coordinates/tests/test_frames.py]: See 757d1d9fd8c8b55ade21f6dc2d566c9ca5e8047f

File: astropy/io/fits/tests/test_table.py
Comment (WilliamJamieson) [inline astropy/io/fits/tests/test_table.py:1242]: @mhvk do you think this is better, or would something akin to your suggestion here https://github.com/astropy/astropy/pull/13754#discussion_r983622791 be better?

Comment (mhvk) [reply astropy/io/fits/tests/test_table.py:1242]: This can just be `np.True_`.

But, yes, I like my suggestion perhaps a little better as it makes clear that the intent of the test is to verify that type of the output from `__eq__` -- and if `numpy` ever where to return python bool scalars, that would be fine!

**Comment by jdavies-st** [comment]
> Pre-commit flake8 checks make sure the `__all__` in `__init__` stays synced with the module content.

Do note that `flake8` will not ensure that `__all__` is correct, only that there are no unused imports.  I.e. it won't report that the following bugs:

```python
__all__ = ['myfunction', 'myclass', myvariable]
```

or

```python
__all__ = {'myfunction', 'myclass', 'myvariable'}
```

`pylint` will find these, but it will also find lots of false positives as well.

Don't ask me how I know this.  😅

File: astropy/table/tests/test_operations.py
Comment (mhvk) [inline astropy/table/tests/test_operations.py:835]: Note the typo: this should check column 'b'

File: astropy/table/tests/test_operations.py
Comment (mhvk) [inline astropy/table/tests/test_operations.py:846]: Same

File: astropy/table/tests/test_operations.py
Comment (mhvk) [inline astropy/table/tests/test_operations.py:861]: Same

File: astropy/table/tests/test_operations.py
Comment (mhvk) [inline astropy/table/tests/test_operations.py:872]: Same

File: astropy/table/tests/test_operations.py
Comment (mhvk) [inline astropy/table/tests/test_operations.py:926]: Same

**Review by mhvk — COMMENTED** [review]
Looks good! I found a few mistaken changes (typos), though.

**Review by mcara — CHANGES_REQUESTED** [review]

**Review by mhvk — APPROVED** [review]
Looks all good! Thanks for sticking with this!

**Comment by mhvk** [comment]
Failure due to a timeout, so unrelated

File: astropy/coordinates/tests/test_frames.py
Comment (saimn) [reply astropy/coordinates/tests/test_frames.py]: While reviewing the black PR for io.fits I see this line 
```
        assert isinstance(v := (tbhdu.columns.columns[4].array[0] == np.True_), (bool, np.bool_)) and v
```

which brings me here through git blame, to ask ... whyyy ?? 
The original code was just testing that an array element was a True value. Why does it need to be so complicated just to avoid a `noqa` flag ?

Comment (WilliamJamieson) [reply astropy/coordinates/tests/test_frames.py]: @saimn, wen I originally "fixed" this, i attempted to change this to `assert tbhdu.columns.columns[4].array[0] is np.True_` (the `noqa` flag was suppressing the suggestion to change this). @mhvk then identified this as a test which was testing the result of the `__eq__` function; meaning that, I needed to change this back. Moreover, @mhvk noted that there are lots of return values from `__eq__` which will technically don't fail the `assert` such as returning `"boo!"`, which means we ought to test that the result is a boolean (the `bool` vs `np.bool_` type was an implementation detail). Hence the result turned into a test of the type of the equality return value.

After thinking about this, the original tests were essentially testing that the array elements were `True` (under the assumption that `__eq__` would be returning a boolean). Which means that the new `isinstance` check would not be able to detect if "True" and "False" were accidentally reversed in the `__eq__` evaluation (or some other pathological bug), which the previous iteration of the test could in fact detect  (assuming the result was a boolean). So I added the walrus assignment to test that the `__eq__` was also the expected outcome in addition to the right type. This was probably overkill, but it means the adjustments to the test should cover all the same cases as before in addition to confirming the boolean nature of the the result.

Feel free to change it back if you would like.

**Pull request merged**
